tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/GenerateTimeline.scala

```diff
@@ -0,0 +1,483 @@
+    // Gather the data
+    val execHostToTaskList = new mutable.TreeMap[String, ArrayBuffer[TimelineTaskInfo]]()
+    val stageIdToColor = mutable.HashMap[Int, String]()
+    var colorIndex = 0
+    var minStartTime = Long.MaxValue
+    var maxEndTime = 0L
+    // Sort the data so
+    app.taskEnd.foreach { tc =>
+      val host = tc.host
+      val execId = tc.executorId
+      val stageId = tc.stageId
+      val taskId = tc.taskId
+      val launchTime = tc.launchTime
+      val finishTime = tc.finishTime
+      val duration = tc.duration
+      val taskInfo = new TimelineTaskInfo(stageId, taskId, launchTime, finishTime, duration)
+      val execHost = s"$execId/$host"
+      execHostToTaskList.getOrElseUpdate(execHost, ArrayBuffer.empty) += taskInfo
+      minStartTime = Math.min(launchTime, minStartTime)
+      maxEndTime = Math.max(finishTime, maxEndTime)
+      stageIdToColor.getOrElseUpdate(stageId, {
+        val color = COLORS(colorIndex % COLORS.length)
+        colorIndex += 1
+        color
+      })
+    }
+
+    val stageRangeInfo = execHostToTaskList.values.flatMap { taskList =>
+      taskList
+    }.groupBy { taskInfo =>
+      taskInfo.stageId
+    }.map {
+      case (stageId, iter) =>
+        val start = iter.map(_.startTime).min
+        val end = iter.map(_.endTime).max
+        new TimelineStageInfo(stageId, start, end, end-start)
+    }
+
+    val stageInfo = app.enhancedStage.map { sc =>
+      val stageId = sc.stageId
+      val submissionTime = sc.submissionTime.get
+      val completionTime = sc.completionTime.get
+      val duration = sc.duration.get
+      minStartTime = Math.min(minStartTime, submissionTime)
+      maxEndTime = Math.max(maxEndTime, completionTime)
+      new TimelineStageInfo(stageId, submissionTime, completionTime, duration)
+    }
+
+    val execHostToSlots = execHostToTaskList.map {
+      case (execHost, taskList) =>
+        (execHost, calcLayoutSlotsNeeded(taskList))
+    }.toMap
+
+    val jobInfo = app.enhancedJob.map { jc =>
+      val jobId = jc.jobID
+      val startTime = jc.startTime
+      val endTime = jc.endTime.get
+      val duration = jc.duration.get
+      minStartTime = Math.min(minStartTime, startTime)
+      maxEndTime = Math.max(maxEndTime, endTime)
+      new TimelineJobInfo(jobId, startTime, endTime, duration)
+    }
+
+    val sqlInfo = app.enhancedSql.map { sc =>
+      val sqlId = sc.sqlID
+      val startTime = sc.startTime
+      val endTime = sc.endTime.get
+      val duration = sc.duration.get
+      minStartTime = Math.min(minStartTime, startTime)
+      maxEndTime = Math.max(maxEndTime, endTime)
+      new TimelineSqlInfo(sqlId, startTime, endTime, duration)
+    }
+
+    // Add 1 second for padding at the end...
+    maxEndTime += 1000
+
+    // Do the high level layout of what the output page should look like
+    // TITLE
+    // EXEC(s)      | TASK TIMING
+    // STAGES       | STAGE TIMING
+    // STAGE RANGES | STAGE RANGE TIMING
```

Comment:
perhaps we should clarify what stage range means vs just stages, from just looking at the output graph it was unclear to me until I read the code